src/components/plugin/offline-storage.ts

```diff
@@ -0,0 +1,88 @@
+import Dexie from "dexie";
+import * as PluginAPI from "@concord-consortium/lara-plugin-api";
+import { ILogEvent, IStudentInfo } from "../../types";
+import { FirestoreBatchUpdateSize, sendBulkLogEventsToFirestore } from "../../db";
+
+export const OfflineStorageTBDMarker = "TBD";
+
+export class DexieStorage extends Dexie {
+  public logs: Dexie.Table<ILogEvent, number>;
+
+  constructor() {
+    super("GlossaryDb");
+    this.version(1).stores({
+      logs: "++id, [resourceUrl+contextId]",   // compound index for speed
+    });
+  }
+}
+
+export const dexieStorage = new DexieStorage();
+
+export const saveLogEventInIndexDB = (logEvent: ILogEvent) => {
+  dexieStorage.logs.put(logEvent);
+};
+
+let syncingLogEvents = false;
+export const syncLogEventsToFirestore = (context: PluginAPI.IPluginRuntimeContext) => {
+  return async (event: PluginAPI.IPluginSyncEvent) => {
+    // don't allow for multiple simultaneous syncs
+    if (syncingLogEvents) {
+      return;
+    }
+    syncingLogEvents = true;
+
+    const {updateCallback} = event;
+    try {
+      updateCallback({status: "started"});
+      if (!studentInfo) {
+        updateCallback({status: "failed", message: "Student info not found!"});
+      } else {
+        const {contextId, userId, source} = studentInfo;
+        const logQuery = dexieStorage.logs
+          .where("[resourceUrl+contextId]")
+          .anyOf([context.resourceUrl, OfflineStorageTBDMarker], [context.resourceUrl, contextId]);
+        let logs = await logQuery.toArray();
+        logs = logs.map(log => ({...log, contextId, userId}));
+
+        const totalLogs = logs.length;
+        updateCallback({status: "working", message: `Found ${totalLogs} logs to sync`});
+
+        if (totalLogs > 0) {
+          // sync in batches
+          let batchIndex = 0;
+          do {
+            const batch = logs.slice(0, FirestoreBatchUpdateSize);
+            logs = logs.slice(FirestoreBatchUpdateSize);
+
+            updateCallback({
+              status: "working",
+              message: `Syncing logs ${batchIndex + 1} - ${batchIndex + batch.length}`
+            });
+            await sendBulkLogEventsToFirestore(source, contextId, batch);
+
+            batchIndex += batch.length;
+          } while (logs.length > 0);
+
+          updateCallback({
+            status: "working",
+            message: `Finished syncing ${totalLogs} logs, deleting locally saved logs`
+          });
+
+          // delete synced logs
+          await logQuery.delete();
```

Comment:
After syncronise review with Noah I will change this to do a query within the inner loop based on the ids and delete them per batch.

Reply:
This has been updated in this commit: https://github.com/concord-consortium/glossary-plugin/pull/84/commits/f24389a82053c1609b394c6f7416796df6987280